### ISSUE TITLE
jobs: panic in jobs metrics poller

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#98730]
Unsure if this is jobs (new metrics poller) or the recent refactor to unregister the metric struct on server shutdown.
@aadityasondhi @andreimatei could you have a look?

Found here https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelEssentialCi/9085880?showRootCauses=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&expandBuildTestsSection=true

```
* ERROR: a panic has occurred!
* runtime error: invalid memory address or nil pointer dereference
* (1) attached stack trace
*   -- stack trace:
*   | runtime.gopanic
*   |   GOROOT/src/runtime/panic.go:884
*   | runtime.panicmem
*   |   GOROOT/src/runtime/panic.go:260
*   | runtime.sigpanic
*   |   GOROOT/src/runtime/signal_unix.go:835
*   | github.com/cockroachdb/cockroach/pkg/jobs.(*ScheduledJob).ExecutorType
*   |   github.com/cockroachdb/cockroach/pkg/jobs/scheduled_job.go:173
*   | github.com/cockroachdb/cockroach/pkg/jobs/metricspoller.processSchedulePTSRecord
*   |   github.com/cockroachdb/cockroach/pkg/jobs/metricspoller/job_statistics.go:227
*   | github.com/cockroachdb/cockroach/pkg/jobs/metricspoller.manageProtectedTimestamps.func1
*   |   github.com/cockroachdb/cockroach/pkg/jobs/metricspoller/job_statistics.go:124
*   | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).Txn.func1
*   |   github.com/cockroachdb/cockroach/pkg/sql/internal.go:1406
*   | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn.func4
*   |   github.com/cockroachdb/cockroach/pkg/sql/internal.go:1484
*   | github.com/cockroachdb/cockroach/pkg/kv.runTxn.func1
*   |   github.com/cockroachdb/cockroach/pkg/kv/db.go:965
*   | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec
*   |   github.com/cockroachdb/cockroach/pkg/kv/txn.go:928
*   | github.com/cockroachdb/cockroach/pkg/kv.runTxn
*   |   github.com/cockroachdb/cockroach/pkg/kv/db.go:964
*   | github.com/cockroachdb/cockroach/pkg/kv.(*DB).TxnWithAdmissionControl
*   |   github.com/cockroachdb/cockroach/pkg/kv/db.go:927
*   | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn
*   |   github.com/cockroachdb/cockroach/pkg/kv/db.go:902
*   | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn
*   |   github.com/cockroachdb/cockroach/pkg/sql/internal.go:1472
*   | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).Txn
*   |   github.com/cockroachdb/cockroach/pkg/sql/internal.go:1407
*   | github.com/cockroachdb/cockroach/pkg/jobs/metricspoller.manageProtectedTimestamps
*   |   github.com/cockroachdb/cockroach/pkg/jobs/metricspoller/job_statistics.go:105
*   | github.com/cockroachdb/cockroach/pkg/jobs/metricspoller.(*metricsPoller).Resume.func1
*   |   github.com/cockroachdb/cockroach/pkg/jobs/metricspoller/poller.go:64
*   | github.com/cockroachdb/cockroach/pkg/jobs/metricspoller.(*metricsPoller).Resume
*   |   github.com/cockroachdb/cockroach/pkg/jobs/metricspoller/poller.go:75
*   | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine.func2
*   |   github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1529
*   | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine
*   |   github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1530
*   | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob
*   |   github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:417
*   | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeJob.func1
*   |   github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:335
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
*   |   github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:470
*   | runtime.goexit
*   |   GOROOT/src/runtime/asm_amd64.s:1594
```

Jira issue: CRDB-25471

**Comment by aadityasondhi** [comment]
Maybe I am missing something but I think the only way it would panic because of the unregistering of the metric struct is if the `NodeMetricsRecorder` is nil. But we do a nil check before adding it to the list of stopper functions: 
https://github.com/cockroachdb/cockroach/pull/98529/files#diff-d58c1926a1ed4d155c9a0360d7b9ed70155e6142cbeb972b96a5a5708137962aR620.

I can take a further look into what other things happen during the stopping of a tenant, but at first look, it doesn't seem that the metric recorder is causing these.

**Comment by abarganier** [comment]
The nil pointer dereference happens here:

https://github.com/cockroachdb/cockroach/blob/ad904b5a9006d613ba5fa338dafb49c8eda6bb1a/pkg/jobs/metricspoller/job_statistics.go#L230

If `sj` is nil, then `sj.ExecutorType()` will panic, which is what's happening here. It's the poller further up the stack that triggers this call, but the fact that the ScheduledJob is nil seems to be an issue with this bit of code returning an unexpected result:

https://github.com/cockroachdb/cockroach/blob/ad904b5a9006d613ba5fa338dafb49c8eda6bb1a/pkg/jobs/metricspoller/job_statistics.go#L222-L223

Assigning over to the disaster recovery team to take a closer look.

**Comment by dt** [comment]
@abarganier wasn't this fixed by #98714?

**Comment by abarganier** [comment]
@dt thanks for linking, I hadn't seen that PR. This indeed looks to solve the issue, so I'll close this out.
